### PR TITLE
Add customized render to chat api as proposed

### PR DESCRIPTION
**Pull request by PorterNan**
Add customized render to chat api as proposed
Clean up some hooks code in composite

# What

# Why

# How Tested

# Process & policy checklist

- [ ] I have updated the project documentation to reflect my changes if necessary.
- [ ] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: packages/react-components/src/components/TypingIndicator.tsx [hunk 1]
```diff
@@ -70,12 +72,12 @@ const getDefaultComponents = (
...
         key={`typing indicator display string ${index}`}
       >
-        {index < typingUsers.length - 1 ? typingUser.displayName + ', ' : typingUser.displayName}
+        {index < typingUsersToRender.length - 1 ? typingUser.displayName + ', ' : typingUser.displayName}
       </span>
     );
   });
 
-  const countOfUsersNotMentioned = typingUsers.length - typingUsersMentioned.length;
+  const countOfUsersNotMentioned = typingUsersToRender.length - typingUsersMentioned.length;
```
Comment (prprabhu-ms) [inline packages/react-components/src/components/TypingIndicator.tsx:80]: Should this continue to use `typingUsers`? It makes sense to not name participants without a `displayName` property but they should be included in the count of number of participants that are typing.

File: packages/react-components/src/components/TypingIndicator.tsx [hunk 2]
```diff
@@ -88,7 +90,8 @@ const getDefaultComponents = (
 };
 
 const defaultTypingString = (typingUsers: CommunicationParticipant[]): string => {
-  return typingUsers.length > 0 ? (typingUsers.length > 1 ? ' are typing...' : ' is typing...') : '';
+  const typingUsersToRender = typingUsers.filter((typingUser) => typingUser.displayName !== undefined);
```
Comment (prprabhu-ms) [inline packages/react-components/src/components/TypingIndicator.tsx:93]: Similar to above, should this continue to use `typingUsers` because participants without a `displayName` are still typing ;)

File: packages/react-composites/src/composites/CallComposite/hooks/usePropsFor.ts [hunk 3]
```diff
@@ -1,72 +1,15 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
 
-import {
```
Comment (prprabhu-ms) [inline packages/react-composites/src/composites/CallComposite/hooks/usePropsFor.ts:4]: Changes to this file seem unrelated to the rest of this PR.

File: packages/react-composites/src/composites/ChatComposite/ChatScreen.tsx [hunk 4]
```diff
@@ -47,6 +57,17 @@ export const ChatScreen = (props: ChatScreenProps): JSX.Element => {
...
   const threadStatusProps = useAdaptedSelector(getThreadStatusProps);
 
+  const onRenderMessageAvatar = useMemo(
+    () => onRenderAvatar && ((userId: string) => onRenderAvatar(userId, 'chatThread')),
+    [onRenderAvatar]
+  );
+
+  const onRenderParticipantAvatar = useMemo(
+    () =>
+      onRenderAvatar && ((participant: CommunicationParticipant) => onRenderAvatar(participant.userId, 'chatThread')),
```
Comment (prprabhu-ms) [inline packages/react-composites/src/composites/ChatComposite/ChatScreen.tsx:67]: s/chatThread/participantList/

File: packages/react-composites/src/composites/ChatComposite/hooks/usePropsFor.ts [hunk 5]
```diff
@@ -1,37 +1,15 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
 
-import { ChatClientState } from 'chat-stateful-client';
```
Comment (prprabhu-ms) [inline packages/react-composites/src/composites/ChatComposite/hooks/usePropsFor.ts:4]: These changes don't seem functionally related to this PR.

**Review by prprabhu-ms — APPROVED** [review]
LGTM, but please remove changes that are unrelated functionally to keep the PR crisp.

**Comment by prprabhu-ms** [comment]
+1 to this not being a `breaking-change`: Adding an optional argument to a function signature is non-breaking in JS. Semantically, the default value for this optional argument is also consistent with previous behaviour.

File: packages/communication-react/review/communication-react.api.md [hunk 6]
```diff
@@ -522,7 +522,9 @@ export type ChatCompositeClientState = {
 export type ChatCompositeProps = {
     adapter: ChatAdapter;
     fluentTheme?: PartialTheme | Theme;
-    onRenderAvatar?: (userId: string) => JSX.Element;
+    onRenderAvatar?: (userId: string, avatarType?: 'chatThread' | 'participantList') => JSX.Element;
```
Comment (JamesBurnside) [inline packages/communication-react/review/communication-react.api.md:525]: this list could get huge (which might be totally fine), but are there other options for what this interface might look like?
e.g.
```
customRenderers?: {
  onRenderAvatar: ()=> JSX.Element,
  onRenderAvatar: ()=> JSX.Element etc.
}
```

(not saying this is better! Just wondering if we explored options for what this customization interface would look like)

Comment (JamesBurnside) [reply packages/communication-react/review/communication-react.api.md:525]: Then would the documentation highlight what the `onRenderAvatar` would change?

Comment (JamesBurnside) [reply packages/communication-react/review/communication-react.api.md:525]: Maybe the interface is a good thing to bring up to the azure review board and get their opinions

**Review by JamesBurnside — APPROVED** [review]
This is awesome! Are we planning to add new custom render functions as we decide/as requested by users? Or is there a plan to cover certain customization scenarios?

File: packages/react-components/src/components/TypingIndicator.tsx (hunk 2 above)
Comment (PorterNan) [reply packages/react-components/src/components/TypingIndicator.tsx:93]: It was filtered outside of this component previously, I just moved the code into the function to make code more clear, it is still sharing the same logic
In previous design, only people with displayname will be shown in the typing indicator like
A, B are typing, this count is used for determine whether it is 'are typing' or 'is typing'

File: packages/communication-react/review/communication-react.api.md (hunk 6 above)
Comment (PorterNan) [reply packages/communication-react/review/communication-react.api.md:525]: Agree, currently this api allows developer to use just one way to render all the avatar(that could be most likely the case), but if they are willing to, they could render different avatar in different places by second parameter
